io.openems.edge.battery.fenecon.home/src/io/openems/edge/battery/fenecon/home/FeneconHomeBatteryImpl.java

```diff
@@ -294,111 +330,286 @@ protected ModbusProtocol defineModbusProtocol() throws OpenemsException {
+		try {
+			// Register is available -> add Registers for current hardware to protocol
+			int offset = ModuleParameters.ADDRESS_OFFSET_FOR_CELL_VOLT_AND_TEMP.getValue();
+			int voltOffset = ModuleParameters.VOLTAGE_ADDRESS_OFFSET.getValue();
+			int voltSensors = ModuleParameters.VOLTAGE_SENSORS_PER_MODULE.getValue();
+			for (int t = 1; t <= numberOfTowers; t++) {
+				String towerString = "TOWER_" + t + "_OFFSET";
+				int towerOffset = ModuleParameters.valueOf(towerString).getValue();
+				for (int i = 1; i < numberOfModulePerTower + 1; i++) {
+					Collection<AbstractModbusElement<?>> elements = new ArrayList<>();
+					for (int j = 0; j < voltSensors; j++) {
+						String key = this.getSingleCellPrefix(j, i, t) + KEY_VOLTAGE;
+						UnsignedWordElement uwe = new UnsignedWordElement(towerOffset + i * offset + voltOffset + j);
+						AbstractModbusElement<?> ame = m(this.channelMap.get(key).channelId(), uwe);
+						elements.add(ame);
+					}
+					this.getModbusProtocol().addTask(new FC3ReadRegistersTask(towerOffset + offset * i + voltOffset,
+							Priority.HIGH, elements.toArray(new AbstractModbusElement<?>[0])));
+				}
+			}
+
+			int tempOffset = ModuleParameters.TEMPERATURE_ADDRESS_OFFSET.getValue();
+			int tempSensors = ModuleParameters.TEMPERATURE_SENSORS_PER_MODULE.getValue();
+			for (int t = 1; t <= numberOfTowers; t++) {
+				String towerString = "TOWER_" + t + "_OFFSET";
+				int towerOffset = ModuleParameters.valueOf(towerString).getValue();
+				for (int i = 1; i < numberOfModulePerTower + 1; i++) {
+					Collection<AbstractModbusElement<?>> elements = new ArrayList<>();
+					for (int j = 0; j < tempSensors; j++) {
+						String key = this.getSingleCellPrefix(j, i, t) + KEY_TEMPERATURE;
+						UnsignedWordElement uwe = new UnsignedWordElement(towerOffset + i * offset + tempOffset + j);
+						AbstractModbusElement<?> ame = m(this.channelMap.get(key).channelId(), uwe);
+						elements.add(ame);
+					}
+					this.getModbusProtocol().addTask(new FC3ReadRegistersTask(towerOffset + offset * i + tempOffset,
+							Priority.HIGH, elements.toArray(new AbstractModbusElement<?>[0])));
+				}
+			}
+		} catch (OpenemsException e) {
+			this.log.info("Cell Temperature and Voltage Dynamic Channels could not created");
+		}
+	}
+
+	/**
+	 * Gets the Number of Modules Per Tower.
+	 * 
+	 * @return the Number of Modules Per Tower as a {@link CompletableFuture}.
+	 * @throws OpenemsException on error
+	 */
+	private CompletableFuture<Integer> getNumberOfModulesPerTowers() {
+		final CompletableFuture<Integer> result = new CompletableFuture<Integer>();
+		try {
+			ModbusUtils.readELementOnce(this.getModbusProtocol(), new UnsignedWordElement(10024), true)
+					.thenAccept(numberOfModulesPerTower -> {
+						if (numberOfModulesPerTower == null) {
+							return;
+						}
+						result.complete(numberOfModulesPerTower);
+					});
+		} catch (OpenemsException e) {
+			this.log.info("Number of modules per tower could not read");
```

Comment:
If for whatever reason this Exception happens, `result` will be never completed.